compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -703,8 +703,7 @@ impl<'tcx> Cx<'tcx> {
                                 tcx,
                                 anon_const.def_id.to_def_id(),
                             )
-                            .instantiate_identity()
```

Comment:
This will break `tests/ui/asm/const-error.rs`, for the record. I also think we should not be obligated to eagerly raise an error for that test, because it's inconsistent with other and only works b/c the const isn't `TooGeneric`.

```
//@ only-x86_64
//@ needs-asm-support

// Test to make sure that we emit const errors eagerly for inline asm

use std::arch::asm;

fn test<T>() {
    unsafe {
        asm!("/* {} */", const 1 / 0);
        //~^ ERROR evaluation of
    }
}

fn main() {}
```

Reply:
c.f.

```
//@ only-x86_64
//@ needs-asm-support

// Test to make sure that we emit const errors eagerly for inline asm

use std::arch::asm;

fn test<T>() {
    unsafe {
        asm!("/* {} */", const 1 / std::mem::size_of::<T>());
    }
}
```

Which obviously only fails if we mono something like `test::<Zst>()`.

Reply:
Ah yeah I just found https://github.com/rust-lang/rust/commit/82f23d56b758d87e5eee642946ad8a422e6542ba. ;)

Reply:
We don't emit const errors eagerly for inline const blocks. So if inline asm behave like those (inheriting the generics from the surrounding function), IMO we should not error eagerly for them either.

Reply:
Cc @Amanieu

Reply:
So right, this brings up a bigger question of "should we try to eagerly eval possibly unreachable but otherwise monomorphic consts"? 

From a technical perspective, this seems like something that *could* be done by a MIR pass if we wanted to do so, though, which calls `const_eval_resolve` on all unevaluated `mir::Const::Unevaluated` and swallows any `TooGeneric`, rather than doing it ad-hoc and within mir build *just* for these consts.

Reply:
We do have such a pass for const items, but those are (currently) always monomorphic.

Reply:
I have adjusted the test, which should make this PR ready to land.